src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,201 @@
+            parsed = self._parse_agent_tag(tag)
+            if parsed is not None:
+                candidate_pairs.append((tag, parsed))
+        candidate_pairs.sort(key=lambda pair: pair[1], reverse=True)
+        candidate_tags = list(map(lambda pair: pair[0], candidate_pairs))
+        previous_tag = None
+        current_digest = self._get_image_digest(current)
+        for candidate_tag in candidate_tags:
+            candidate_digest = self._get_image_digest(f"{repo}:{candidate_tag}")
+            if candidate_digest != current_digest:
+                previous_tag = candidate_tag
+                break
+        if previous_tag is None:
+            return None
+        return f"{repo}:{previous_tag}"
+
+    def _get_broken_agent_image(self, current: str) -> str:
+        """Returns an image reference that doesn't exist."""
+        repo = self._get_image_repo(current)
+        return f"{repo}:broken"
+
+    @JunitTestSuite()
+    def test_upgrade_agent(self, cluster, namespace, k8s_client):
+        """
+        This test will prepare the cluster with the previous agent image. Once it is ready to install it will restart
+        the service with an agent image that doesn't exist, so that all the hosts will move to `insufficient`. Then it
+        will restart the service again with the current image, so that host will upgrade and move to `ready` again.
+        """
+        current_image = self._get_current_agent_image(k8s_client, namespace)
+        previous_image = self._get_previous_agent_image(current_image)
+        broken_image = self._get_broken_agent_image(current_image)
+        assert current_image
+        assert previous_image
+        assert broken_image
+        log.info(f"Previous image is '{previous_image}'")
+        log.info(f"Broken image is '{broken_image}'")
+        log.info(f"Current image is '{current_image}'")
+
+        try:
+            # Prepare the cluster for installation using the previous image:
+            log.info("Waiting cluster to move to 'ready' with previous image")
+            self._load_service_with_agent_image(k8s_client, namespace, previous_image)
+            cluster.prepare_for_installation()
+            cluster.wait_for_ready_to_install()
+
+            # Restart the service with the broken agent image, so that nodes will not be able to pull it and move to the
+            # insufficient state:
+            log.info("Waiting for hosts to move to 'insufficient' with broken image")
+            self._load_service_with_agent_image(k8s_client, namespace, broken_image)
+            cluster.wait_until_hosts_are_insufficient()
+
+            # Restart the agent with the current agent image and wait for all hosts to be ready, which means that they
+            # upgraded correctly:
+            log.info("Waiting for hosts to move to 'ready'")
+            self._load_service_with_agent_image(k8s_client, namespace, current_image)
+            cluster.wait_for_ready_to_install()
+        finally:
```

Comment:
Not sure I understood why the try-finally is here, van you elaborate?

Reply:
The reason for this is to restore the service configuration if the test fails in the middle.